app/javascript/dashboard/components/widgets/WootWriter/Editor.vue

```diff
@@ -618,6 +607,44 @@ export default {
     onFocus() {
       this.$emit('focus');
     },
+    insertContentIntoEditor(content, defaultFrom = 0) {
+      const from = defaultFrom || this.editorView.state.selection.from || 0;
+      let node = new MessageMarkdownTransformer(messageSchema).parse(content);
+
+      this.insertNodeIntoEditor(node, from, undefined, content);
+    },
+    insertNodeIntoEditor(node, from = 0, to = 0, content = '') {
+      if (!this.editorView) {
+        return;
+      }
+
+      // If the node is of type 'doc' and has only one child which is a paragraph,
+      // then extract its inline content to be inserted as inline.
+      if (
+        node.type.name === 'doc' &&
+        node.childCount === 1 &&
+        node.firstChild.type.name === 'paragraph'
+      ) {
+        node = node.firstChild.content;
+      } else if (content && node.textContent === content) {
+        node = this.editorView.state.schema.text(content);
+      }
+
+      let tr;
+      if (to) {
+        tr = this.editorView.state.tr
+          .replaceWith(from, to, node)
+          .insertText(` `);
+      } else {
+        tr = this.editorView.state.tr.insert(from, node);
+      }
+
+      this.state = this.editorView.state.apply(tr);
```

Comment:
Would it be a good idea to move this part into a function outside the vue file, to a util? I am think that could make it easy to write unit tests. It can simply return the transaction to be applied

Reply:
Agreed, this piece is hard to understand. It would be good to accompany it with relevant tests.